btree/btree.c

```diff
@@ -1247,6 +1249,14 @@ struct m0_btree_oimpl {
 	/** Node descriptor for cookie if it is going to be used. **/
 	struct nd      *i_cookie_node;
 
+	/**
+	 * Array to hold addresses of node descriptors which are going to be
+	 * captured in transaction. This will be used by transaction engine for
+	 * callbacks on transaction commit. Array size is the worst case based
+	 * on btree_callback_credit() calculation.
+	 */
+	void           *i_callback[BTREE_CALLBACK_CREDIT];
```

Comment:
Do not assign the max array size here. This will not work if the current transaction is going to capture multiple trees and each of them is currently of MAX_HEIGHT. Instead the call to the credit calculator needs to be used for every tree.
Also the type of this should not be void * - instead it should be some structure which contains 2 memberes: callback function pointer and callback data pointer,

Reply:
Agree with the below point
```
Also the type of this should not be void * - instead it should be some structure which contains 2 memberes: callback function pointer and callback data pointer
```
For below point, this structure is in `m0_btree_oimpl`, so it will be called for each operation. And for multiple btrees credit calculator will also needs to be called for each btree. I will discuss with you offline.
```
Do not assign the max array size here. This will not work if the current transaction is going to capture multiple trees and each of them is currently of MAX_HEIGHT. Instead the call to the credit calculator needs to be used for every tree.
```